amethyst_utils/src/auto_fov.rs

```diff
@@ -0,0 +1,248 @@
+        AutoFov {
+            base_aspect_ratio: (horizontal, vertical),
+            ..self
+        }
+    }
+
+    /// Builds a new instance with the given `min_fovx` value
+    ///
+    /// This function panics if the given `min_fovx` is not larger than zero or is larger than
+    /// `max_fovx`.
+    pub fn with_min(self, min: f32) -> Self {
+        let max = self.max_fovx;
+        self.with_min_max(min, max)
+    }
+
+    /// Builds a new instance with the given `max_fovx` value
+    ///
+    /// This function panics if the given `max_fovx` is smaller than `min_fovx`.
+    pub fn with_max(self, max: f32) -> Self {
+        let min = self.min_fovx;
+        self.with_min_max(min, max)
+    }
+
+    /// Builds a new instance with the given `min_fovx` and `max_fovx` values
+    ///
+    /// This function panics if the given `min_fovx` is not larger than zero, or if the given
+    /// `max_fovx` is smaller than the given `min_fovx`.
+    pub fn with_min_max(self, min: f32, max: f32) -> Self {
+        assert!(
+            min > 0.0,
+            format!("`min_fovx` should be larger than 0, but `{}` given", min)
+        );
+        assert!(
+            max >= min,
+            format!(
+                "`max_fovx` should be larger than or equal to `min_fovx` which is `{}`, but `{}` given",
+                min,
+                max,
+            ),
+        );
+        AutoFov {
+            min_fovx: min,
+            max_fovx: max,
+            ..self
+        }
+    }
+
+    /// Computes the new horizontal FOV from the current screen aspect ratio and vertical FOV
+    pub fn new_fovx(&self, current_aspect_ratio: f32, fovy: f32) -> f32 {
+        let delta_aspect = current_aspect_ratio - self.base_aspect_value();
+
+        if delta_aspect.abs() <= ::std::f32::EPSILON {
+            return self.base_fovx;
+        }
+
+        let fovy = if self.fixed_grow_rate { 1.0 } else { fovy };
+        let delta_fovx = self.fovx_grow_rate * fovy * delta_aspect;
+        let new_fovx = self.base_fovx + delta_fovx;
+
+        new_fovx.max(self.min_fovx).min(self.max_fovx)
+    }
+
+    #[inline]
+    fn base_aspect_value(&self) -> f32 {
+        self.base_aspect_ratio.0 as f32 / self.base_aspect_ratio.1 as f32
+    }
+}
+
+impl Component for AutoFov {
+    type Storage = HashMapStorage<Self>;
+}
+
+impl<'a> PrefabData<'a> for AutoFov {
+    type SystemData = WriteStorage<'a, AutoFov>;
+    type Result = ();
+
+    fn add_to_entity(
+        &self,
+        entity: Entity,
+        storage: &mut Self::SystemData,
+        _: &[Entity],
+    ) -> Result<Self::Result, PrefabError> {
+        storage.insert(entity, self.clone())?;
+        Ok(())
+    }
+}
+
+impl Default for AutoFov {
+    fn default() -> Self {
+        AutoFov {
+            base_fovx: 1.861684535,
+            fovx_grow_rate: 1.0,
+            fixed_grow_rate: false,
+            base_aspect_ratio: (16, 9),
+            min_fovx: 0.1,
+            max_fovx: ::std::f32::consts::PI,
```

Comment:
The leading `::` are not needed in 2018 edition iirc

Reply:
```suggestion
            max_fovx: std::f32::consts::PI,
```